cpp/src/arrow/io/file.h

```diff
@@ -183,10 +183,12 @@ class ARROW_EXPORT MemoryMappedFile : public ReadWriteFileInterface {
 
   /// Create new file with indicated size, return in read/write mode
   static Status Create(const std::string& path, int64_t size,
-                       std::shared_ptr<MemoryMappedFile>* out);
+                       std::shared_ptr<MemoryMappedFile>* out,
+		       const off_t offset = 0);
 
   static Status Open(const std::string& path, FileMode::type mode,
-                     std::shared_ptr<MemoryMappedFile>* out);
+                     std::shared_ptr<MemoryMappedFile>* out,
+		     const off_t offset = 0);
```

Comment:
I think it's better to use `int64_t` for the `off_t` is a platform-specific type. 

I also agree with Ben about adding an overload `Open(path, mode, offset, &mmap)`

Reply:
Hi @wesm  we may need to add another param `length` here, with this user could memory map a portion of the file just like the semantics of `mmap()`. 

- MemoryMappedFile::Open(path, mode, &mmap)
- MemoryMappedFile::Open(path, mode, length, offset, &mmap)

In the 2nd case, since only part of the file is mmaped, I think we may need to make `MemoryMappedFile::GetSize() ` return `length` instead of the `size_` of the corresponding file. 
Does this looks like the right approach to you?